File: brainfuck_parser.js
function brainfuck_parser(code,input) {

  var cells = [];
  var cellIndex = 0;
  var toPrint = '';
  var loopStart = [];  
  var loopEnd = [];
  var inputs = 0;

  for (i = 0; i < code.length ; i++) {

    //console.log("cells: " , i, cells)
    //console.log("i: " , i)

    if (code.charAt(i) === '+') {
      if(cells[cellIndex] != undefined) {
        var item = cells[cellIndex];
        cells.splice(cellIndex,1);
        item++;
        cells.splice(cellIndex,0,item);
      }
      else cells.push(1);
      //console.log("+++: ", cells[cellIndex])
    }

    else if (code.charAt(i) === '-') {
      if(cells[cellIndex] != undefined) {
        var item = cells[cellIndex];
        cells.splice(cellIndex,1);
        item--;
        cells.splice(cellIndex,0,item);
      }
      else cells.push(1);
      //console.log("---: " , cells[cellIndex])
    }

    else if (code.charAt(i) === '>') {
      cellIndex++;
      //console.log(">>>: " , cellIndex)
    }

    else if (code.charAt(i) === '<') {
      cellIndex--;
      //console.log("<<<: " , cellIndex)
    }

    else if (code.charAt(i) === '[') {
      if (cells[cellIndex] === 0) {
        i = loopEnd.pop();
      }
      else {
        loopStart.push(i);
        loopEnd.pop();
      }
      //console.log("[[[: " , loopStart)
    }
  
    else if (code.charAt(i) === ']') {
      loopEnd.push(i);
      i = loopStart.pop() - 1;
      //console.log("]]]: " , loopEnd)
    }

    else if (code.charAt(i) === '.') {
      toPrint += String.fromCharCode(cells[cellIndex]);
      //console.log("toPrint: " , toPrint)
    }

    else if (code.charAt(i) === ',') {
      cells[cellIndex] = input.charCodeAt(inputs);
      inputs++;
    }
  }
  return toPrint;
}

var Paula = "++++++++[>+++[>+++>++++>+++++>+++++<<<<-]>+>>>--[<]<-]>>.>+.>---.>++++.<<.";
var Paaba = "++++++++[>+++[>+++>++++>+++++>+++++<<<<-]>+>>>--[<]<-]>>.>+.>---,.>,.<.";
var HelloWorld = "++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.";

console.log(brainfuck_parser(Paula))
console.log(brainfuck_parser(Paaba,'ab'))
console.log(brainfuck_parser(HelloWorld))

File: brainfuck_parser.pl
#!/usr/bin/perl

sub brainfuck_parser {

  my ($code, $input) = @_;
  my $cells = ();
  my $cellIndex = 0;
  my $toPrint = '';
  my $loopStart = ();
  my $loopEnd = ();
  my $inputs = 0;

  for (my $i = 0; $i < length $code; $i++)
    {
    
    #print "cells: @cells\n";
    
    if ( substr($code,$i,1) =~ /\+/ )
    {
      if ( $cells[$cellIndex] != null )
      {
        my $item = $cells[$cellIndex];
        $item++;
        $cells[$cellIndex] = $item;
      }
      else
      {
        $cells[$cellIndex] = 1;
      }
    #print "+++ $cells[$cellIndex]\n";
    }
    
    elsif ( substr($code,$i,1) =~ /\-/ )
    {
      if ( $cells[$cellIndex] != null )
      {
        my $item = $cells[$cellIndex];
        $item--;
        $cells[$cellIndex] = $item;
      }
      else
      {
        $cells[$cellIndex] = 1;
      }
    #print "--- $cells[$cellIndex]\n";
    }
    
    elsif ( substr($code,$i,1) =~ />/ )
    {
      $cellIndex++;
      #print ">>> $i $cellIndex\n";
    }
    
    elsif ( substr($code,$i,1) =~ /</ )
    {
      $cellIndex--;
      #print "<<< $i $cellIndex\n";
    }
    
    elsif ( substr($code,$i,1) =~ /\[/ )
    {
      if ($cells[$cellIndex] == 0 )
      {
        $i = pop(@loopEnd);
      }
      else
      {
        push(@loopStart,$i);
        pop(@loopEnd);
      }
    #print "[[[ @loopStart\n";
    }
    
    elsif ( substr($code,$i,1) =~ /\]/ )
    {
      push(@loopEnd,$i);
      $i = pop(@loopStart) - 1;
      #print "]]] @loopEnd\n";
    }
    
    elsif ( substr($code,$i,1) =~ /\./ )
    {
      $toPrint .= chr($cells[$cellIndex]);
      #print "... $i $cells[$cellIndex] $toPrint\n";
    }
    
    elsif ( substr($code,$i,1) =~ /\,/ )
    {
      my $temppp = substr($input,$inputs,1);
      $cells[$cellIndex] = ord($temppp);
      $inputs++;
      #print ",,, $i $toPrint\n";
    }
  }
  
  return $toPrint;

}

#my $Paula = "++++++++[>+++[>+++>++++>+++++>+++++<<<<-]>+>>>--[<]<-]>>.>+.>---.>++++.<<.";
#my $result = brainfuck_parser($Paula);
#print "$result\n";

#my $Paaba = "++++++++[>+++[>+++>++++>+++++>+++++<<<<-]>+>>>--[<]<-]>>.>+.>---,.>,.<.";
#my $result2 = brainfuck_parser($Paaba,"ab");
#print "$result2\n";

my $HelloWorld =
"++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.";
my $result3 = brainfuck_parser($HelloWorld);
print "$result3\n";

